### PR TITLE
PerformanceObserver should not throw when entryTypes is empty

### LINKED ISSUE #7: PerformanceObserver should not throw when entryTypes is empty
Right now polyfill throws an error when there is no valid entry types in a call to `observe()`: https://github.com/fastly/performance-observer-polyfill/blob/master/src/observer.ts#L12

The latest spec version does not say anymore that observer should throw an error - rather it should issue a console warning

> Remove all types from entry types that are not contained in relevantGlobal's frozen array of supported entry types. The user agent SHOULD notify developers if entry types is modified. For example, a console warning listing removed types might be appropriate.
If the resulting entry types sequence is an empty sequence, abort these steps. The user agent SHOULD notify developers when the steps are aborted to notify that registration has been aborted. For example, a console warning might be appropriate.

https://w3c.github.io/performance-timeline/#observe-method

Explanation why it shouldn't throw is given in https://github.com/w3c/performance-timeline/issues/87.

Also, latest web platform tests could work as a reference: https://github.com/web-platform-tests/wpt/blob/master/performance-timeline/po-observe.any.js#L24-L35

### PR DESCRIPTION
**Pull request by phamann**
Fixes: #7 

### TL;DR
The [Performance Timeline Level 2](https://w3c.github.io/performance-timeline/) spec has changed since we originally wrote the polyfill (with some additions and behaviour changes). One change is to no longer throw if no valid entry types are provided in a call to `observe()` - a discussion as to why is documented in https://github.com/w3c/performance-timeline/issues/87. This PR attempts to ensure the library comply with the spec changes. 

### Notes:
- We now throw only if neither `entryTypes` or `type` is supplied, or if BOTH are supplied. 
- We warn if invalid entry types are supplied via a `console.warn` message.
- We abort with a warning if no valid entry types are found.

The test suite has been changed to reflect the above changes.

### CODE REVIEW

**Comment by crellison** [comment]
This looks good to me. Thanks for addressing the issue so promptly!

File: src/observer.spec.ts [hunk 1]
```diff
@@ -69,15 +90,27 @@ describe("PerformanceObserver", (): void => {
...
+    it("should warn if invaild entryTypes are supplied", (): void => {
+      const observer = new PerformanceObserver(
+        callbackFixture,
+        new PerformanceObserverTaskQueue()
+      );
+      observer.observe({ entryTypes: ["resource", "mark", "bad"] });
+      expect(consoleOutput).toHaveLength(1);
+    });
+
+    it("should abort if no vaild entryTypes are supplied", (): void => {
```
Comment (victor-homyakov) [inline src/observer.spec.ts]: Typo: "vaild" -> "valid"

File: src/observer.spec.ts [hunk 2]
```diff
@@ -69,15 +90,27 @@ describe("PerformanceObserver", (): void => {
       expect(observer.entryTypes).toEqual(["resource", "mark"]);
     });
 
-    it("should throw if no vaild entryTypes are supplied", (): void => {
+    it("should warn if invaild entryTypes are supplied", (): void => {
```
Comment (victor-homyakov) [inline src/observer.spec.ts]: Typo: "invaild" -> "invalid"

File: src/observer.ts [hunk 3]
```diff
@@ -40,14 +44,31 @@ class PerformanceObserver implements PollingPerformanceObserver {
...
       throw new Error(ERRORS["no-entry-types"]);
     }
 
-    const entryTypes: string[] = options.entryTypes.filter(isValidType);
+    if (options.entryTypes && options.type) {
+      throw new Error(ERRORS["both-entry-types"]);
+    }
+
+    let rawEntryTypes: string[] = [];
+    if (options && options.entryTypes) {
```
Comment (victor-homyakov) [inline src/observer.ts]: Redundant check for `if (options)`: `options` cannot be undefined or null at this point. Could be simplified to `if (options.entryTypes)`.

File: src/observer.ts [hunk 4]
```diff
@@ -40,14 +44,31 @@ class PerformanceObserver implements PollingPerformanceObserver {
...
+      rawEntryTypes = options.entryTypes;
+    }
+    if (options && options.type) {
```
Comment (victor-homyakov) [inline src/observer.ts]: This line also could be simplified to `if (options.type) {`

File: src/observer.ts [hunk 5]
```diff
@@ -40,14 +44,31 @@ class PerformanceObserver implements PollingPerformanceObserver {
...
+    if (!options || (!options.entryTypes && !options.type)) {
       throw new Error(ERRORS["no-entry-types"]);
     }
 
-    const entryTypes: string[] = options.entryTypes.filter(isValidType);
+    if (options.entryTypes && options.type) {
+      throw new Error(ERRORS["both-entry-types"]);
+    }
+
+    let rawEntryTypes: string[] = [];
```
Comment (victor-homyakov) [inline src/observer.ts]: Could this code fragment be rewritten? This way we can eliminate redundant array initialization `rawEntryTypes: string[] = []`

```ts
  public observe(options?: PerformanceObserverInit): void {
    if (!options) {
      throw new Error(ERRORS["no-entry-types"]);
    }

    let rawEntryTypes: string[];
    if (options.entryTypes && options.type) {
      throw new Error(ERRORS["both-entry-types"]);
    } else if (options.entryTypes) {
      rawEntryTypes = options.entryTypes;
    } else if (options.type) {
      rawEntryTypes = [options.type];
    } else {
      throw new Error(ERRORS["no-entry-types"]);
    }
```

**Review by victor-homyakov — CHANGES_REQUESTED** [review]

File: src/observer.ts (hunk 5 above)
Comment (phamann) [reply src/observer.ts]: Have cleaned this up in https://github.com/fastly/performance-observer-polyfill/commit/6d7f6c77b4eae56ce970682839657ba4af07035f

**Review by doramatadora — APPROVED** [review]
With the offline clarification that we are happy with the risk of not type checking the `entryTypes` array (Chrome throws a runtime error, as per spec): I think this looks good! Thank you @phamann 🙌
